fix: Missing types in "mysql" import

wellwelwel opened the pull request:
I performed a variety of tests to cover the most common options from `mysql` import among:

```ts
.promise();
/* callbacks */

.query(sql);
.query(sql, params);

.execute(sql);
.execute(sql, params);
```

Included in:

```ts
.createConnection();
.createPool();
.createPool().getConnection();
```

<hr />

### This fixes these problems:

1. In `Pool` the `execute` method didn't exists (#1966)
2. In `Connection` all the `execute` method params and returns were set as `any`
3. In `promise()` the most of options were missing, but they didn't report errors because it were implied as `any` [(as in this comment)](https://github.com/sidorares/node-mysql2/issues/1936#issuecomment-1506258067)
4. Fix type conflicts between `promise()` and `callback` use cases

<hr />

#### These changes don't affects the `mysql/promise`

<hr />

### Tests:

-  The changes are small, but tests are extensive, so I preferred to embed them in **details** tag.
-  All `.promise()` tests the `query` and `execute` methods with and without **Prepared Statements**
-  These fixes includes all variations of the following tests

<hr />

<details>
<summary>
   Base connection
</summary>

```ts
import mysql from 'mysql2';

const access: mysql.ConnectionOptions = {
   host: '127.0.0.1',
   user: 'root',
   password: '*',
   port: 3306,
   database: 'mydb',
};

const sqlQuery = 'SELECT * FROM `Users`;';
const sqlQueryWithParam = 'SELECT * FROM `Users` WHERE `id` = ?;';
const sqlParam = [1];
```

</details>

<hr />

<details>
<summary>
   <code>.createConnection()</code>, <code>.promise()</code> and <code>.query()</code>
</summary>

```ts
(async () => {
   const db = mysql.createConnection(access).promise();
   const [users] = await db.query(sqlQuery);
   const [usersWithParam] = await db.query(sqlQueryWithParam, sqlParam);

   console.log(users, usersWithParam);
   await db.end();
})();
```

</details>

<details>
<summary>
   <code>.createPool()</code>, <code>.promise()</code> and <code>.query()</code>
</summary>

```ts
(async () => {
   const db = mysql.createPool(access).promise();
   const [users] = await db.query(sqlQuery);
   const [usersWithParam] = await db.query(sqlQueryWithParam, sqlParam);

   console.log(users, usersWithParam);
   await db.end();
})();
```

</details>

<details>
<summary>
   <code>.createPool()</code>, <code>.promise()</code>, <code>.getConnection()</code> and <code>.query()</code>
</summary>

```ts
(async () => {
   const db = mysql.createPool(access).promise();
   const [users] = await db.query(sqlQuery);
   const [usersWithParam] = await db.query(sqlQueryWithParam, sqlParam);

   console.log(users, usersWithParam);
   await db.end();
})();
```

</details>

<hr />

<details>
<summary>
   <code>.createConnection()</code>, <code>.promise()</code> and <code>.execute()</code>
</summary>

```ts
(async () => {
   const db = mysql.createConnection(access).promise();

   const [users] = await db.execute(sqlQuery);
   const [usersWithParam] = await db.execute(sqlQueryWithParam, sqlParam);

   console.log(users, usersWithParam);
   await db.end();
})();
```

</details>

<details>
<summary>
   <code>.createPool()</code>, <code>.promise()</code> and <code>.execute()</code>
</summary>

```ts
(async () => {
   const db = mysql.createPool(access).promise();
   const [users] = await db.execute(sqlQuery);
   const [usersWithParam] = await db.execute(sqlQueryWithParam, sqlParam);

   console.log(users, usersWithParam);
   await db.end();
})();
})();
```

</details>

<details>
<summary>
   <code>.createPool()</code>, <code>.promise()</code>, <code>.getConnection()</code> and <code>.execute()</code>
</summary>

```ts
(async () => {
   const db = mysql.createPool(access).promise();
   const conn = await db.getConnection();
   const [users] = await conn.execute(sqlQuery);
   const [usersWithParam] = await conn.query(sqlQueryWithParam, sqlParam);

   console.log(users, usersWithParam);
   await db.end();
})();
```

</details>

<hr />

<details>
<summary>
   <strong>callback:</strong> <code>.createConnection()</code> and <code>.query()</code>
</summary>

```ts
(async () => {
   const db = mysql.createConnection(access);

   db.query(sqlQuery, (error, users) => {
      console.log(users);
      db.end();
   });
})();
```

</details>

<details>
<summary>
   <strong>callback:</strong> <code>.createConnection()</code> and <code>.query()</code> with Prepared Statements
</summary>

```ts
(async () => {
   const db = mysql.createConnection(access);

   db.query(sqlQueryWithParam, sqlParam, (error, usersWithParam) => {
      console.log(usersWithParam);
      db.end();
   });
})();
```

</details>

<hr />

<details>
<summary>
   <strong>callback:</strong> <code>.createConnection()</code> and <code>.execute()</code>
</summary>

```ts
(async () => {
   const db = mysql.createConnection(access);

   db.execute(sqlQuery, (error, users) => {
      console.log(users);
      db.end();
   });
})();
```

</details>

<details>
<summary>
   <strong>callback:</strong> <code>.createConnection()</code> and <code>.execute()</code> with Prepared Statements
</summary>

```ts
(async () => {
   const db = mysql.createConnection(access);

   db.execute(sqlQueryWithParam, sqlParam, (error, usersWithParam) => {
      console.log(usersWithParam);
      db.end();
   });
})();
```

</details>

<hr />

<details>
<summary>
   <strong>callback:</strong> <code>.createPool()</code> and <code>.query()</code>
</summary>

```ts
(async () => {
   const db = mysql.createPool(access);

   db.query(sqlQuery, (error, users) => {
      console.log(users);
      db.end();
   });
})();
```

</details>

<details>
<summary>
   <strong>callback:</strong> <code>.createPool()</code> and <code>.query()</code> with Prepared Statements
</summary>

```ts
(async () => {
   const db = mysql.createPool(access);

   db.query(sqlQueryWithParam, sqlParam, (error, usersWithParam) => {
      console.log(usersWithParam);
      db.end();
   });
})();
```

</details>

<hr />

<details>
<summary>
   <strong>callback:</strong> <code>.createPool()</code>, <code>.getConnection()</code> and <code>.query()</code>
</summary>

```ts
(async () => {
   const db = mysql.createPool(access);

   db.getConnection((err, conn) => {
      conn.query(sqlQuery, (error, users) => {
         console.log(users);
         db.end();
      });
   });
})();
```

</details>

<details>
<summary>
  <strong>callback:</strong> <code>.createPool()</code>, <code>.getConnection()</code> and <code>.query()</code> with Prepared Statements
</summary>

```ts
(async () => {
   const db = mysql.createPool(access);

   db.getConnection((err, conn) => {
      conn.query(sqlQueryWithParam, sqlParam, (error, usersWithParam) => {
         console.log(usersWithParam);
         db.end();
      });
   });
})();
```

</details>

<hr />

<details>
<summary>
   <strong>callback:</strong> <code>.createPool()</code> and <code>.execute()</code>
</summary>

```ts
(async () => {
   const db = mysql.createPool(access);

   db.execute(sqlQuery, (error, users) => {
      console.log(users);
      db.end();
   });
})();
```

</details>

<details>
<summary>
   <strong>callback:</strong> <code>.createPool()</code> and <code>.execute()</code> with Prepared Statements
</summary>

```ts
(async () => {
   const db = mysql.createPool(access);

   db.execute(sqlQueryWithParam, sqlParam, (error, usersWithParam) => {
      console.log(usersWithParam);
      db.end();
   });
})();
```

</details>

<hr />

<details>
<summary>
   <strong>callback:</strong> <code>.createPool()</code>, <code>.getConnection()</code> and <code>.execute()</code>
</summary>

```ts
(async () => {
   const db = mysql.createPool(access);

   db.getConnection((err, conn) => {
      conn.execute(sqlQuery, (error, users) => {
         console.log(users);
         db.end();
      });
   });
})();
```

</details>

<details>
<summary>
  <strong>callback:</strong> <code>.createPool()</code>, <code>.getConnection()</code> and <code>.execute()</code> with Prepared Statements
</summary>

```ts
(async () => {
   const db = mysql.createPool(access);

   db.getConnection((err, conn) => {
      conn.execute(sqlQueryWithParam, sqlParam, (error, usersWithParam) => {
         console.log(usersWithParam);
         db.end();
      });
   });
})();
```

</details>

sidorares commented:
Not sure I understand "The changes are small, but tests are extensive, so I preferred to embed them in details tag."
Can we add some typing tests to cover the fix?

wellwelwel commented:
@sidorares When I referred to the changes being small, this was in relation to the fixed types.

About the tests being extensive, I was referring to the number of possibilities between asynchronous (async/await) and callbacks combinations varying between the `query` and `execute` methods inside the `createPool`, `createConnection` and `getConnection`.

I do this to ensure that the changes cover all possible variations.

Would you like me to create a test that checks all these types automatically in the future? I can attach this test to the TypeScript CI.

wellwelwel commented:
@sidorares Can I append this coverage tests to **TypeScript Build CI**?

This would be interesting, given the number of combinations between the methods that have `query` and `execute`.

And about "details tag", it's just the markdown formatting "\<details>" 🙋🏻‍♂️

> I deleted my previous comment

wellwelwel commented:
@sidorares

I have automated all the tests and implemented it to the "TypeScript Build Test".
I also reworked all the PR description.

Can you review it? 🙋🏻‍♂️

wellwelwel commented:
### Before this PR

I would like to better explain how this test works:

-  It handles the imports "mysql" and "mysql/promise" as external
-  It get all the contents based in declared types from [index.d.ts](https://github.com/sidorares/node-mysql2/blob/master/index.d.ts) and [promise.d.ts](https://github.com/sidorares/node-mysql2/blob/master/promise.d.ts)
-  Then, it try to build the project and if succeed, it means that all types are compatible, but is important note:
   -  Compatible types don't mean correct types, it preventing only changes that break the TypeScript compilation in projects that imports `mysql2` or `mysql2/promise`.

<hr />

### After this PR

Through this new way, now it's possible to simule real use codes from every content in `mysql` and `mysql2/promise`.  
To do this, simply simulate a basic real code that uses the method you want, by example:

In `createPool`, it has the `promise` method from `PromisePoolConnection`. So I can use:

```ts
const conn = await mysql.createPool(access).promise();
const pool = await conn.getConnection(); // getConnection method exists in PromisePoolConnection ✅
```

Suppose someone changes the `Pool` typings and the promise goes to `PromiseConnection`, now it will break the TypeScript build:

```ts
const conn = await mysql.createPool(access).promise();
const pool = await conn.getConnection(); // getConnection method doesn't exists in PromiseConnection ❌
```

With this, the `TypeScript Build Test` will prevent a change that breaks the functioning of any type that the tests have already created, in addition to the overall test compilation.

<hr />

That's basically it and as before, if you have any question, just let me know 🙋🏻‍♂️

sidorares commented:
Sorry for the delay in reviewing - I'm relatively inexperienced in TS, don't want to merge in a rush, also hoping some other eyes can see the change

It looks good though

wellwelwel commented:
@sidorares It's fine 🙋🏻‍♂️

So, if you want, I can rollback the automated tests and keep only the type fixes.

After that, instead of a PR, I can open a discussion about the automated type tests. What do you think?

marcellino-ornelas commented:
@sidorares I took a look at some of the typescript changes and they look good to me. 

After merged and deployed I can pull down in my project to see if it fixes https://github.com/sidorares/node-mysql2/issues/1816

Type changes to execute look good to me although Im not using pool so wont be able to verify that myself

however the tests @wellwelwel looks good and seems like they would cover most situations

wellwelwel commented:
> After merged and deployed I can pull down in my project to see if it fixes https://github.com/sidorares/node-mysql2/issues/1816

@marcellino-ornelas, you can test it without the merge/deploy by cloning locally the master from this PR, then:

- Install it locally in a test project: `npm i ./my_path/node-mysql2`
- Try to run your build 🤹🏻

sidorares commented:
Thanks for 👀 @marcellino-ornelas!

sidorares commented:
Probably this https://www.githubstatus.com/incidents/nf7s6933tnn8
